presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/util/Memoizers.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore.util;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+public final class Memoizers
+{
+    private Memoizers()
+    {
+    }
+
+    public static final class Simple<T>
```

Comment:
I think it would be cleaner to make these methods that return standard functional types
```java
public static Function<T, T> memoizeLast() ...

public static Function<I, O> memoizeLast(Function<I, O> transform) ...
```
Rather than exposing the implementation classes in the usages

Reply:
Done, except using `UnaryOperator<T>` for `memoizeLast()` to reduce the verbosity of type signatures.